### PR TITLE
Dev 3

### PR DESCRIPTION
**Pull request by derins**
# Current issues
none

# Code changes
1. data collection code created

# Documentation changes
1. readme added to powerflex_data branch

# Validation suggestions
none

### CODE REVIEW

**Review by jongoncalves — COMMENTED** [review]
Instead of making comments on the individual lines I'm just going make them all here at once:
- Remove the gitignore changes - we definitely don't want to ignore a file called config
- All API requests should be functions
- Fix indentation on your comments
- All string concatenation need to be `f` strings
- Follow proper PEP8 formatting, e.g. spaces after colons on key value pairs
- UPPER_SNAKE_CASE what you think need to be constants
- Remove commented out code
- Provide instructions on how to run this code. The readme shows `python data_collection.py` but never tells the user/developer which dependencies exist or that he needs to install any.
- Your usage of double quotes and single quotes are inconsistent. Double quote everything!
- Why are you grabbing session data from 10 days ago? Explain with a comment in the code.
- Saving data to s3 should be a separate modular function.
- Make this python 3.7+ compliant 
- Username and password should be parameters passed in

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# username / password
+config.py
+./powerflex_data/config.py
```
Comment (jongoncalves) [inline .gitignore]: Why is this still gitignored?

File: powerflex_data/data_collection.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    token_meas = json.loads(resp_login.text)["access_token"]
+    auth_token_meas = "Bearer " + token_meas
+    headers_meas = {"cache-control": "no-cache", "content-type":
+                    "application/json", "Authorization": auth_token_meas}
+
+    token_arch = json.loads(resp_archive.text)["access_token"]
+    auth_token_arch = "Bearer " + token_arch
+    headers = {"cache-control": "no-cache", "content-type": "application/json",
+               "Authorization": auth_token_arch}
+    return (headers_meas, headers)
```
Comment (jongoncalves) [inline powerflex_data/data_collection.py]: Two login requests in one function that do the exact same thing. DRY this up and make it generic.

File: powerflex_data/data_collection.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def interval_data_collection(start, end, headers_meas):
+    data_meas = json.dumps({"measurement": "ct_response",
+                            "time_filter": [start, end]})
+    r = requests.post(URL_MEASUREMENT, headers=headers_meas, data=data_meas)
+    data = json.loads(r.text)
```
Comment (jongoncalves) [inline powerflex_data/data_collection.py]: Same comment for all json loads throughout, but why can't you just use the builtin `requests` `.json()`?

File: powerflex_data/data_collection.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                            [start, end]})
+    data_in = json.dumps({"startTime": start, "stopTime": end, "filterBy":
+                          "sessionStopTime", "anonymize": True})
+    r1 = requests.post(URL_ARCH1, headers=headers, data=data_in)
+    r2 = requests.post(URL_ARCH2, headers=headers, data=data_in)
+    data1 = json.loads(r1.text)
+    df1 = pd.DataFrame(data1["sessions"])
+    data2 = json.loads(r2.text)
+    df2 = pd.DataFrame(data2["sessions"])
+    return df1, df2
```
Comment (jongoncalves) [inline powerflex_data/data_collection.py]: Same thing as the login. You are repeating your self here. DRY up this logic and make it more generic.

File: powerflex_data/data_collection.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+# urls for session data from two different sites
+URL_ARCH1 = "https://archive.powerflex.com/get_csh/0032/01"
+URL_ARCH2 = "https://archive.powerflex.com/get_csh/0032/02"
+
+URL_LOGIN = "https://slac.powerflex.com:9443/login"
+URL_ARCHIVE = "https://archive.powerflex.com/login"
+
+def api_login(username, password):
+    data_login = json.dumps({"username": username, "password": password})
+    headers = {'cache-control': 'no-cache', 'content-type': 'application/json'}
```
Comment (jongoncalves) [inline powerflex_data/data_collection.py]: Still single quotes....

File: powerflex_data/data_collection.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
+import s3fs
+import boto3
+import pandas as pd
+import requests
+import json
+import time
+import datetime
+import sys
+from io import StringIO
+import boto3
```
Comment (jongoncalves) [inline powerflex_data/data_collection.py]: why is boto3 imported twice?

File: powerflex_data/environment.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  - s3transfer=0.3.3=py38_0
+  - setuptools=45.2.0=py38_0
+  - six=1.14.0=py38_0
+  - sqlite=3.31.1=ha441bb4_0
+  - tk=8.6.8=ha441bb4_0
+  - urllib3=1.25.8=py38_0
+  - wheel=0.34.2=py38_0
+  - xz=5.2.4=h1de35cc_4
+  - zlib=1.2.11=h1de35cc_3
+prefix: /Users/derins/miniconda3/envs/powerflex_data_env
```
Comment (jongoncalves) [inline powerflex_data/environment.yml]: This yml needs revamped. Please remove your personal path from the prefix. And remove all dependencies listed other than the ones required by the project. Even though conda adds additional ones. Don't list them here for Devine.

File: .gitignore (hunk 1 above)
Comment (derins) [reply .gitignore]: Fixed

File: powerflex_data/data_collection.py (hunk 2 above)
Comment (derins) [reply powerflex_data/data_collection.py]: Done.

File: powerflex_data/data_collection.py (hunk 3 above)
Comment (derins) [reply powerflex_data/data_collection.py]: Done.

File: powerflex_data/data_collection.py (hunk 4 above)
Comment (derins) [reply powerflex_data/data_collection.py]: Done.

File: powerflex_data/data_collection.py (hunk 5 above)
Comment (derins) [reply powerflex_data/data_collection.py]: Fixed.

File: powerflex_data/data_collection.py (hunk 6 above)
Comment (derins) [reply powerflex_data/data_collection.py]: Fixed.

File: powerflex_data/environment.yml (hunk 7 above)
Comment (derins) [reply powerflex_data/environment.yml]: Unused dependencies and prefix is removed.

**Review by jongoncalves — APPROVED** [review]

**Pull request merged**